packages/tpetra/core/src/Tpetra_Experimental_BlockCrsMatrix_def.hpp

```diff
@@ -2835,91 +2925,92 @@ class GetLocalDiagCopy {
+      Kokkos::DualView<const local_ordinal_type*, device_type> exportLIDsTemp = exportLIDs;
+      exportLIDsTemp.sync_host();
+    }
+    auto exportLIDsHost = exportLIDs.view_host();
+    auto numPacketsPerLIDHost = numPacketsPerLID.view_host(); // we will modify this
+    {
+      typedef Impl::BlockCrsReducer<Impl::BlockCrsRowStruct<size_t>,host_exec> reducer_type;
+      const auto policy = Kokkos::RangePolicy<host_exec>(0, numExportLIDs);
+      Kokkos::parallel_reduce
+        (policy, 
+         [=](const int &i, typename reducer_type::value_type &update) {
+          const LO lclRow = exportLIDsHost(i);
+          size_t numEnt = srcGraph.getNumEntriesInLocalRow (lclRow);
+          numEnt = (numEnt == Teuchos::OrdinalTraits<size_t>::invalid () ? 0 : numEnt);
+          
+          const size_t numBytes = packRowCount<LO, GO, host_exec> (numEnt, numBytesPerValue, blockSize);
+          numPacketsPerLIDHost(i) = numBytes;
+          update += typename reducer_type::value_type(numEnt, numBytes, numEnt);
+        }, rowReducerStruct);
+      {
+        Kokkos::DualView<size_t*, buffer_device_type> numPacketsPerLIDTemp = numPacketsPerLID;
+        numPacketsPerLIDTemp.modify_host();
       }
-
-      const size_t numBytes =
-        packRowCount<LO, GO, HES> (numEnt, numBytesPerValue, blockSize);
-      numPacketsPerLID[i] = numBytes;
-      totalNumBytes += numBytes;
-      totalNumEntries += numEnt;
-      maxRowLength = std::max (maxRowLength, numEnt);
     }
-
-    if (debug) {
-      const int myRank = graph_.getComm ()->getRank ();
+    
+    // Compute the number of bytes ("packets") per row to pack.  While
+    // we're at it, compute the total # of block entries to send, and
+    // the max # of block entries in any of the rows we're sending.
+    const size_t totalNumBytes   = rowReducerStruct.totalNumBytes;
+    const size_t totalNumEntries = rowReducerStruct.totalNumEntries;
+    const size_t maxRowLength    = rowReducerStruct.maxRowLength;
+    
+    if (verbose) {
       std::ostringstream os;
-      os << "Proc " << myRank << ": packAndPrepare: totalNumBytes = "
-         << totalNumBytes << endl;
+      os << prefix 
+         << "totalNumBytes = " << totalNumBytes << ", totalNumEntries = " << totalNumEntries 
+         << std::endl;
       std::cerr << os.str ();
     }
 
     // We use a "struct of arrays" approach to packing each row's
     // entries.  All the column indices (as global indices) go first,
     // then all their owning process ranks, and then the values.
-    exports.resize (totalNumBytes);
+    exports.resize (totalNumBytes/numBytesPerValue);
     if (totalNumEntries > 0) {
-      View<char*, HES, MemoryUnmanaged> exportsK (exports.getRawPtr (),
-                                                  totalNumBytes);
+      // exports is resized in the above and we assume the data on device is invalidated. 
+      Kokkos::View<char*,host_exec> exportsByteHost ((char*)exports.view_host().data(), totalNumBytes);
```

Comment:
In the future, prefer `reinterpret_cast` to C-style cast, as it's easier to search for the former.